scripts/ISM/ISM.py

```diff
@@ -0,0 +1,263 @@
+
+This script pulls down teh controls in a 'dumb' way from the xml to get the actual controls. A full featured catalog
+will need to parse appropriate word / xml documents to provide groups /guidance.
+"""
+import io
+import json
+import logging
+import pathlib
+import sys
+import urllib.request
+import zipfile
+from datetime import datetime
+from uuid import uuid4
+
+from ilcli import Command
+
+import trestle.oscal.catalog as catalog
+import trestle.oscal.common as common
+import trestle.oscal.profile as profile
+
+import xmltodict
+
+# Globally define logging behaviour.
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+logger.addHandler(logging.StreamHandler(sys.stdout))
+
+remarks_tuple = '\n'.join(
+    [
+        'This is not an official version of the Australian Government Information Security Manual.',
+        '',
+        'Find the official versions here: https://www.cyber.gov.au/acsc/view-all-content/ism',
+        'This content was generated using scrips/ISM/ISM.py'
+    ]
+)
+
+
+class ISMManager():
+    """ISMManager a class to manage conversion of ISM artifacts into OSCAL."""
+
+    def __init__(self):
+        """Initialize ISM manager. No required parameters."""
+        self._profile_controls = {'OFFICIAL': [], 'PROTECTED': [], 'SECRET': [], 'TOP_SECRET': []}
+        self._profiles = {}
+
+    def fetch_ism(self, url):
+        """Fetch an Australian government ISM and covert to a dict."""
+        logger.debug('Fetching ISM from: ' + url)
+        request_url = urllib.request.urlopen(url)
+        document = request_url.read()
+        zipfile_content = zipfile.ZipFile(io.BytesIO(document))
+        content_list = zipfile_content.namelist()
+        xml_files = [x for x in content_list if '.xml' in x]
+        assert len(xml_files) == 1
+        self.ism_xml = xmltodict.parse(zipfile_content.open(xml_files[0]).read())
+
+    def _populate_control_list(self, control, raw_id):
+        """Populate control lists based on a dict from the xml version of the ISM."""
+        # TODO: Really not pythonic but anyway.
+        control_id = 'control-' + raw_id
+
+        for security_level in self._profile_controls.keys():
+            # Dealing with schema changes 'Yes' and 'true' appear to both be valid options.
+            if control[security_level].lower() == 'yes' or control[security_level].lower() == 'true':
+                self._profile_controls[security_level].append(control_id)
+
+    def _probe_for_keys(self, ism_control):
+        """Probe for the appropriate keys for l2 groups based on whether or not section exists."""
+        l2_group_key = 'Section'
+
+        if l2_group_key not in ism_control.keys():
+            l2_group_key = 'Topic'
+        return l2_group_key
+
+    def _name_clean(self, name: str) -> str:
+        """Normalize string to ncname format."""
+        return input.strip().lower().replace(' ', '_').replace('/', '-')
```

Comment:
looks like bug - input should be name?

Reply:
Thanks I changed it when committing the demo (after testing).